File: apps/mac-app/Sources/Insta360Sync/Insta360/Insta360MediaProto.swift
import Foundation

struct MediaFileEntry: Equatable, Sendable {
    var sourcePath: String
    var size: Int64?
    var captureTime: Int64?
}

enum Insta360MediaProto {
    private static let captureTimePattern = #"^(?:VID|IMG|LRV)_(\d{4})(\d{2})(\d{2})_(\d{6})"#

    static func captureTimeFromFilename(_ name: String) -> Int64? {
        guard let regex = try? NSRegularExpression(pattern: captureTimePattern, options: .caseInsensitive),
              let match = regex.firstMatch(in: name, range: NSRange(name.startIndex..., in: name)),
              match.numberOfRanges >= 5,
              let yRange = Range(match.range(at: 1), in: name),
              let mRange = Range(match.range(at: 2), in: name),
              let dRange = Range(match.range(at: 3), in: name),
              let tRange = Range(match.range(at: 4), in: name) else {
            return nil
        }
        let digits = "\(name[yRange])\(name[mRange])\(name[dRange])\(name[tRange])"
        return Int64(digits)
    }

    static func parseMediaFileEntries(from data: Data) -> [MediaFileEntry] {
        let paths = Insta360Paths.parseMediaPaths(from: data)
        let bytes = [UInt8](data)
        return paths.map { path in
            var captureTime: Int64?
            var size: Int64?
            let pathBytes = Array(path.utf8)
            guard !pathBytes.isEmpty, !bytes.isEmpty else {
                let name = (path as NSString).lastPathComponent
                return MediaFileEntry(
                    sourcePath: path,
                    size: nil,
                    captureTime: captureTimeFromFilename(name)
                )
            }

            var search = 0
            while search <= bytes.count - pathBytes.count {
                guard let idx = indexOf(pathBytes, in: bytes, startingAt: search) else { break }
                if let meta = metaAtPath(bytes: bytes, pathOffset: idx, pathLength: pathBytes.count) {
                    if let ct = meta.captureTime { captureTime = ct }
                    if let sz = meta.size { size = sz }
                    if captureTime != nil || size != nil { break }
                }
                search = idx + 1
            }
            let name = (path as NSString).lastPathComponent
            if captureTime == nil {
                captureTime = captureTimeFromFilename(name)
            }
            return MediaFileEntry(sourcePath: path, size: size, captureTime: captureTime)
        }
    }

    private static func metaAtPath(bytes: [UInt8], pathOffset: Int, pathLength: Int) -> MediaFileEntry? {
        guard pathOffset >= 0,
              pathLength > 0,
              pathOffset + pathLength <= bytes.count else {
            return nil
        }
        let start = max(0, pathOffset - 12)
        for tagPos in start..<pathOffset {
            guard tagPos < bytes.count, bytes[tagPos] == 0x0A else { continue }
            guard tagPos + 1 < bytes.count,
                  let strLen = readVarint(bytes, offset: tagPos + 1) else { continue }
            let (length, strStart) = strLen
            guard strStart == pathOffset, length == pathLength else { continue }
            return parseFileEntry(Array(bytes[tagPos...]))
        }
        return nil
    }

    private static func parseFileEntry(_ entry: [UInt8]) -> MediaFileEntry? {
        var offset = 0
        var captureTime: Int64?
        var size: Int64?
        while offset < entry.count {
            guard let keyPair = readVarint(entry, offset: offset) else { break }
            let (key, next) = keyPair
            offset = next
            let field = Int(key >> 3)
            let wire = Int(key & 0x07)
            switch wire {
            case 0:
                guard let valuePair = readVarint(entry, offset: offset) else { return nil }
                offset = valuePair.1
            case 2:
                guard let lenPair = readVarint(entry, offset: offset) else { return nil }
                let (length, lenNext) = lenPair
                offset = lenNext
                guard offset + length <= entry.count else { return nil }
                if field == 1 {
                    offset += length
                } else if field == 2 {
                    let meta = Array(entry[offset..<(offset + length)])
                    offset += length
                    if let nested = parseNestedMeta(meta) {
                        if let ct = nested.captureTime { captureTime = ct }
                        if let sz = nested.size { size = sz }
                    }
                    return MediaFileEntry(sourcePath: "", size: size, captureTime: captureTime)
                } else {
                    offset += length
                }
            case 1:
                offset += 8
            case 5:
                offset += 4
            default:
                return MediaFileEntry(sourcePath: "", size: size, captureTime: captureTime)
            }
        }
        return MediaFileEntry(sourcePath: "", size: size, captureTime: captureTime)
    }

    private static func parseNestedMeta(_ meta: [UInt8]) -> MediaFileEntry? {
        var offset = 0
        var captureTime: Int64?
        var size: Int64?
        while offset < meta.count {
            guard let keyPair = readVarint(meta, offset: offset) else { break }
            let (key, next) = keyPair
            offset = next
            let field = Int(key >> 3)
            let wire = Int(key & 0x07)
            switch wire {
            case 0:
                guard let valuePair = readVarint(meta, offset: offset) else { return nil }
                let (value, valueNext) = valuePair
                offset = valueNext
                if field == 7 { captureTime = Int64(value) }
                if field == 9 { size = Int64(value) }
            case 2:
                guard let lenPair = readVarint(meta, offset: offset) else { return nil }
                let (length, lenNext) = lenPair
                offset = lenNext + length
            case 1:
                offset += 8
            case 5:
                offset += 4
            default:
                return MediaFileEntry(sourcePath: "", size: size, captureTime: captureTime)
            }
        }
        return MediaFileEntry(sourcePath: "", size: size, captureTime: captureTime)
    }

    private static func indexOf(_ needle: [UInt8], in haystack: [UInt8], startingAt start: Int) -> Int? {
        guard !needle.isEmpty, start >= 0, start <= haystack.count - needle.count else { return nil }
        let limit = haystack.count - needle.count
        for index in start...limit {
            if haystack[index..<(index + needle.count)].elementsEqual(needle) {
                return index
            }
        }
        return nil
    }

    private static func readVarint(_ bytes: [UInt8], offset: Int) -> (Int, Int)? {
        var result = 0
        var shift = 0
        var index = offset
        while index < bytes.count {
            let byte = Int(bytes[index])
            index += 1
            result |= (byte & 0x7F) << shift
            if (byte & 0x80) == 0 {
                return (result, index)
            }
            shift += 7
            if shift > 63 { return nil }
        }
        return nil
    }
}

File: scripts/insta360_media_proto.py
"""Parse Insta360 Luna Ultra UCD2 file-list protobuf entries."""

from __future__ import annotations

import re
from dataclasses import dataclass

from insta360_paths import dedupe_storage_paths, parse_media_paths

CAPTURE_TIME_RE = re.compile(
    r"^(?:VID|IMG|LRV)_(\d{4})(\d{2})(\d{2})_(\d{6})",
    re.IGNORECASE,
)


@dataclass(frozen=True)
class MediaFileEntry:
    source_path: str
    size: int | None = None
    capture_time: int | None = None


def capture_time_from_filename(name: str) -> int | None:
    match = CAPTURE_TIME_RE.match(name)
    if not match:
        return None
    y, mo, d, t = match.groups()
    return int(f"{y}{mo}{d}{t}")


def _read_varint(data: bytes, offset: int) -> tuple[int, int]:
    result = 0
    shift = 0
    while offset < len(data):
        byte = data[offset]
        offset += 1
        result |= (byte & 0x7F) << shift
        if not (byte & 0x80):
            return result, offset
        shift += 7
        if shift > 63:
            raise ValueError("invalid varint")
    raise ValueError("unexpected EOF while reading varint")


def _skip_field(data: bytes, wire_type: int, offset: int) -> int:
    if wire_type == 0:
        _, offset = _read_varint(data, offset)
        return offset
    if wire_type == 2:
        length, offset = _read_varint(data, offset)
        return offset + length
    if wire_type == 1:
        return offset + 8
    if wire_type == 5:
        return offset + 4
    raise ValueError(f"unsupported wire type {wire_type}")


def _parse_nested_meta(meta: bytes) -> tuple[int | None, int | None]:
    capture_time: int | None = None
    size: int | None = None
    offset = 0
    while offset < len(meta):
        key, offset = _read_varint(meta, offset)
        field = key >> 3
        wire = key & 7
        if wire == 0:
            value, offset = _read_varint(meta, offset)
            if field == 7:
                capture_time = value
            elif field == 9:
                size = value
        else:
            offset = _skip_field(meta, wire, offset)
    return capture_time, size


def _parse_file_entry(entry: bytes) -> tuple[int | None, int | None]:
    capture_time: int | None = None
    size: int | None = None
    offset = 0
    while offset < len(entry):
        key, offset = _read_varint(entry, offset)
        field = key >> 3
        wire = key & 7
        if field == 1 and wire == 2:
            length, offset = _read_varint(entry, offset)
            offset += length
        elif field == 2 and wire == 2:
            length, offset = _read_varint(entry, offset)
            meta = entry[offset : offset + length]
            ct, sz = _parse_nested_meta(meta)
            if ct is not None:
                capture_time = ct
            if sz is not None:
                size = sz
            break
        else:
            try:
                offset = _skip_field(entry, wire, offset)
            except ValueError:
                break
    return capture_time, size


def _meta_at_path(data: bytes, path_offset: int, path_len: int) -> tuple[int | None, int | None]:
    for tag_pos in range(max(0, path_offset - 12), path_offset):
        if data[tag_pos] != 0x0A:
            continue
        try:
            str_len, str_start = _read_varint(data, tag_pos + 1)
        except ValueError:
            continue
        if str_start == path_offset and str_len == path_len:
            try:
                return _parse_file_entry(data[tag_pos:])
            except ValueError:
                return None, None
    return None, None


def parse_media_file_entries(data: bytes) -> list[MediaFileEntry]:
    """Return media paths with UCD2 protobuf metadata (field 7=capture time, field 9=size)."""
    paths = dedupe_storage_paths(parse_media_paths(data))
    entries: list[MediaFileEntry] = []
    for path in paths:
        path_bytes = path.encode("ascii")
        capture_time: int | None = None
        size: int | None = None
        search = 0
        while search >= 0:
            idx = data.find(path_bytes, search)
            if idx < 0:
                break
            ct, sz = _meta_at_path(data, idx, len(path_bytes))
            if ct is not None:
                capture_time = ct
            if sz is not None:
                size = sz
            if capture_time is not None or size is not None:
                break
            search = idx + 1
        name = path.rsplit("/", 1)[-1]
        if capture_time is None:
            capture_time = capture_time_from_filename(name)
        entries.append(
            MediaFileEntry(
                source_path=path,
                size=size,
                capture_time=capture_time,
            )
        )
    return entries
